[Quant] [PT2] Enable QLinear input with multi dims

leslie-fang-intel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113928
* #113912
* __->__ #113733

**Summary**
In the previous QLinear implementation, it was assumed that inputs have a dimension of 2. In this update, we have modified QLinear to accept inputs with a dimension greater than 2, incorporating input and output reshaping accordingly.

**Test Plan**
```
python -u -m pytest -s -v test_quantized_op.py -k test_qlinear_pt2e
```


cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @voznesenskym @penguinwu @EikanWang @Guobing-Chen @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

leslie-fang-intel committed:
Enable QLinear input with multi dims

[ghstack-poisoned]

leslie-fang-intel commented:
cc @jianan-gu

jgong5 commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -931,10 +931,19 @@ static at::Tensor linear_int8_with_onednn_weight(
+  // If the input has more than two dimensions, we will reshape it to a 2-dimensional form
+  // for calculation and subsequently reshape the output back.
+  auto input_contig =
+      dim == 2 ? input.contiguous() : input.reshape({-1, input.size(dim - 1)});
```
I don't think `reshape` can guarantee the contiguity of the result. Suggest to explicitly make the result contiguous. A separate question: can we support transposed input?

leslie-fang-intel replied:
> I don't think reshape can guarantee the contiguity of the result. Suggest to explicitly make the result contiguous.

Thanks for the comments and changed.

leslie-fang-intel replied:
> A separate question: can we support transposed input?

I think basically it should depend on whether oneDNN can support the calculation with transposed input, we may check with oneDNN team for that. And for now, in the inductor lowering, the BF16/INT8 Linear assumes the input and output buffer are contiguous as in: https://github.com/pytorch/pytorch/blob/8241fe6edb11278a236e61f4a87eae6be9f032d9/torch/_inductor/ir.py#L5190 and https://github.com/pytorch/pytorch/blob/8241fe6edb11278a236e61f4a87eae6be9f032d9/torch/_inductor/ir.py#L4825-L4828

So, it may be ok if the backend only support contiguous input and output.

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QLinear input with multi dims"


**Summary**
Previously QLinear implementation assumes inputs is with dim of 2. In this diff, we make QLinear accepts input of dim more than 2 with input and output reshape.

**Test Plan**
```
python -u -m pytest -s -v test_quantized_op.py -k test_qlinear_pt2e
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/quantized/cpu/qlinear.cpp:
```diff
@@ -934,7 +934,7 @@ static at::Tensor linear_int8_with_onednn_weight(
   auto input_contig =
-      dim == 2 ? input.contiguous() : input.reshape({-1, input.size(dim - 1)});
+      dim == 2 ? input.contiguous() : input.reshape({-1, input.size(dim - 1)}).contiguous();
   TORCH_CHECK(input_contig.is_contiguous(), "Expect qlinear input to be contiguous.");
```
nit: this check seems not necessary.

leslie-fang-intel replied:
Thanks for the comment and removed.

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QLinear input with multi dims"


**Summary**
In the previous QLinear implementation, it was assumed that inputs have a dimension of 2. In this update, we have modified QLinear to accept inputs with a dimension greater than 2, incorporating input and output reshaping accordingly.

**Test Plan**
```
python -u -m pytest -s -v test_quantized_op.py -k test_qlinear_pt2e
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QLinear input with multi dims"


**Summary**
In the previous QLinear implementation, it was assumed that inputs have a dimension of 2. In this update, we have modified QLinear to accept inputs with a dimension greater than 2, incorporating input and output reshaping accordingly.

**Test Plan**
```
python -u -m pytest -s -v test_quantized_op.py -k test_qlinear_pt2e
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
Hi @jerryzh168, could you kindly help to take a look of this PR?

leslie-fang-intel committed:
Update on "[Quant] [PT2] Enable QLinear input with multi dims"


**Summary**
In the previous QLinear implementation, it was assumed that inputs have a dimension of 2. In this update, we have modified QLinear to accept inputs with a dimension greater than 2, incorporating input and output reshaping accordingly.

**Test Plan**
```
python -u -m pytest -s -v test_quantized_op.py -k test_qlinear_pt2e
```


cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10 voznesenskym penguinwu EikanWang Guobing-Chen zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

leslie-fang-intel commented:
@pytorchbot merge
